Tools/scripts/freeze_modules.py

```diff
@@ -0,0 +1,499 @@
+        lines = infile.readlines()
+        lines = replace_block(
+            lines,
+            "FROZEN_FILES =",
+            "# End FROZEN_FILES",
+            frozenfiles,
+            MAKEFILE,
+        )
+        lines = replace_block(
+            lines,
+            "# BEGIN: freezing modules",
+            "# END: freezing modules",
+            rules,
+            MAKEFILE,
+        )
+        outfile.writelines(lines)
+
+
+def regen_pcbuild(frozenids, frozen):
+    projlines = []
+    filterlines = []
+    for frozenid in frozenids:
+        pyfile, frozenfile = frozen[frozenid]
+
+        _pyfile = os.path.relpath(pyfile, ROOT_DIR).replace('/', '\\')
+        header = os.path.relpath(frozenfile, ROOT_DIR).replace('/', '\\')
+        intfile = header.split('\\')[-1].strip('.h') + '.g.h'
+        projlines.append(f'    <None Include="..\\{_pyfile}">')
+        projlines.append(f'      <ModName>{frozenid}</ModName>')
+        projlines.append(f'      <IntFile>$(IntDir){intfile}</IntFile>')
+        projlines.append(f'      <OutFile>$(PySourcePath){header}</OutFile>')
+        projlines.append(f'    </None>')
+
+        filterlines.append(f'    <None Include="..\\{_pyfile}">')
+        filterlines.append('      <Filter>Python Files</Filter>')
+        filterlines.append('    </None>')
+
+    with updating_file_with_tmpfile(PCBUILD_PROJECT) as (infile, outfile):
+        lines = infile.readlines()
+        lines = replace_block(
+            lines,
+            '<!-- BEGIN frozen modules -->',
+            '<!-- END frozen modules -->',
+            projlines,
+            PCBUILD_PROJECT,
+        )
+        outfile.writelines(lines)
+    with updating_file_with_tmpfile(PCBUILD_FILTERS) as (infile, outfile):
+        lines = infile.readlines()
+        lines = replace_block(
+            lines,
+            '<!-- BEGIN frozen modules -->',
+            '<!-- END frozen modules -->',
+            filterlines,
+            PCBUILD_FILTERS,
+        )
+        outfile.writelines(lines)
+
+
+#######################################
+# freezing modules
+
+def freeze_module(modname, pyfile=None, destdir=MODULES_DIR):
+    """Generate the frozen module .h file for the given module."""
+    for modname, pyfile, ispkg in resolve_modules(modname, pyfile):
+        frozenfile = _resolve_frozen(modname, destdir)
+        _freeze_module(modname, pyfile, frozenfile)
+
+
+def _freeze_module(frozenid, pyfile, frozenfile):
+    tmpfile = frozenfile + '.new'
+
+    argv = [TOOL, frozenid, pyfile, tmpfile]
+    print('#', ' '.join(argv))
+    subprocess.run(argv, check=True)
+
+    os.replace(tmpfile, frozenfile)
+
+
+#######################################
+# the script
+
+def parse_args(argv=sys.argv[1:], prog=sys.argv[0]):
+    import argparse
+    parser = argparse.ArgumentParser(prog=prog)
+    parser.add_argument('--no-regen', dest='regen', action='store_false')
+    parser.add_argument('--no-freeze', dest='freeze', action='store_false')
```

Comment:
Do we need these flags? Why not always regenerate and never freeze (since the Makefile will freeze)?

Reply:
Yeah, they're not needed anymore.  So we can drop the CLI.